#! /bin/sh /usr/share/dpatch/dpatch-run
## opt-54_deltimeshiftrec.dpatch by Tobias Grimm <tg@e-tobi.net>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Offers the choice to delete timeshift recordings
## DP: Taken from the Zulu extensions patch
## DP: See http://www.zulu-entertainment.de/download.php?group=VDR

@DPATCH@
diff -urNad vdr-1.6.0~/config.c vdr-1.6.0/config.c
--- vdr-1.6.0~/config.c	2008-07-25 23:55:15.000000000 +0200
+++ vdr-1.6.0/config.c	2008-07-25 23:57:18.000000000 +0200
@@ -276,6 +276,7 @@
   TimeTransponder = 0;
   MarginStart = 2;
   MarginStop = 10;
+  DelTimeshiftRec = 0;
   AudioLanguages[0] = -1;
   DisplaySubtitles = 0;
   SubtitleLanguages[0] = -1;
@@ -522,6 +523,7 @@
   else if (!strcasecmp(Name, "TimeTransponder"))     TimeTransponder    = atoi(Value);
   else if (!strcasecmp(Name, "MarginStart"))         MarginStart        = atoi(Value);
   else if (!strcasecmp(Name, "MarginStop"))          MarginStop         = atoi(Value);
+  else if (!strcasecmp(Name, "DelTimeshiftRec"))     DelTimeshiftRec    = atoi(Value);
   else if (!strcasecmp(Name, "AudioLanguages"))      return ParseLanguages(Value, AudioLanguages);
   else if (!strcasecmp(Name, "DisplaySubtitles"))    DisplaySubtitles   = atoi(Value);
   else if (!strcasecmp(Name, "SubtitleLanguages"))   return ParseLanguages(Value, SubtitleLanguages);
@@ -632,6 +634,7 @@
   Store("TimeTransponder",    TimeTransponder);
   Store("MarginStart",        MarginStart);
   Store("MarginStop",         MarginStop);
+  Store("DelTimeshiftRec",    DelTimeshiftRec);
   StoreLanguages("AudioLanguages", AudioLanguages);
   Store("DisplaySubtitles",   DisplaySubtitles);
   StoreLanguages("SubtitleLanguages", SubtitleLanguages);
diff -urNad vdr-1.6.0~/config.h vdr-1.6.0/config.h
--- vdr-1.6.0~/config.h	2008-07-25 23:55:15.000000000 +0200
+++ vdr-1.6.0/config.h	2008-07-25 23:57:56.000000000 +0200
@@ -253,6 +253,7 @@
   int TimeSource;
   int TimeTransponder;
   int MarginStart, MarginStop;
+  int DelTimeshiftRec;
   int AudioLanguages[I18N_MAX_LANGUAGES + 1];
   int DisplaySubtitles;
   int SubtitleLanguages[I18N_MAX_LANGUAGES + 1];
diff -urNad vdr-1.6.0~/menu.c vdr-1.6.0/menu.c
--- vdr-1.6.0~/menu.c	2008-07-25 23:55:15.000000000 +0200
+++ vdr-1.6.0/menu.c	2008-07-25 23:58:27.000000000 +0200
@@ -3170,12 +3170,17 @@
 // --- cMenuSetupRecord ------------------------------------------------------
 
 class cMenuSetupRecord : public cMenuSetupBase {
+private:
+  const char *DelTimeshiftRecValues[3];
 public:
   cMenuSetupRecord(void);
   };
 
 cMenuSetupRecord::cMenuSetupRecord(void)
 {
+  DelTimeshiftRecValues[0] = tr("request");
+  DelTimeshiftRecValues[1] = tr("no");
+  DelTimeshiftRecValues[2] = tr("yes");
   SetSection(tr("Recording"));
   Add(new cMenuEditIntItem( tr("Setup.Recording$Margin at start (min)"),     &data.MarginStart));
   Add(new cMenuEditIntItem( tr("Setup.Recording$Margin at stop (min)"),      &data.MarginStop));
@@ -3198,6 +3203,7 @@
   Add(new cMenuEditBoolItem(tr("Setup.Recording$Show time"),                 &data.ShowRecTime));
   Add(new cMenuEditBoolItem(tr("Setup.Recording$Show length"),               &data.ShowRecLength));
   Add(new cMenuEditBoolItem(tr("Setup.CutTimePatch$Adapt start time to cutting marks"), &data.CutTimePatchEnabled));
+  Add(new cMenuEditStraItem(tr("Setup.Recording$Delete timeshift recording"), &data.DelTimeshiftRec, 3, DelTimeshiftRecValues));
 }
 
 // --- cMenuSetupReplay ------------------------------------------------------
@@ -4666,6 +4672,56 @@
      currentReplayControl = NULL;
 }
 
+void cReplayControl::Stop(void)
+{
+  int dummy;
+  bool playing = GetIndex(dummy, dummy, false);
+  cRecordControl* rc = cRecordControls::GetRecordControl(fileName);
+
+  if (playing && rc && rc->InstantId()) {
+     isyslog("found Timeshiftrecording");
+
+     if ((Setup.DelTimeshiftRec != 0 ) || (Interface->Confirm(tr("Delete recording?")))) {
+        cRecordControl *rc = cRecordControls::GetRecordControl(fileName);
+        if (rc) {
+           cTimer *timer = rc->Timer();
+           if (timer) {
+              const char* reccmd_backup = cRecordingUserCommand::GetCommand();
+              cRecordingUserCommand::SetCommand(NULL);
+ 
+              timer->Skip();
+              cRecordControls::Process(time(NULL));
+              if (timer->IsSingleEvent()) {
+                 isyslog("deleting timer %s", *timer->ToDescr());
+                 Timers.Del(timer);
+                 }
+              Timers.SetModified();
+
+              // restore reccmd
+              cRecordingUserCommand::SetCommand(reccmd_backup);
+              }
+           }
+        isyslog("stop replaying %s", fileName);
+        cDvbPlayerControl::Stop();
+
+        if (Setup.DelTimeshiftRec != 1) {
+           cRecording *recording = Recordings.GetByName(fileName);;
+           if (recording) {
+              if (recording->Delete()) {
+                 Recordings.DelByName(fileName);
+                 ClearLastReplayed(fileName);
+                 return;
+                 }
+              else
+                 Skins.Message(mtError, tr("Error while deleting recording!"));
+              }
+           }
+        }
+     else
+        cDvbPlayerControl::Stop();
+     }
+}
+
 void cReplayControl::SetRecording(const char *FileName, const char *Title)
 {
   free(fileName);
@@ -5080,7 +5136,12 @@
                            else
                               Show();
                            break;
-            case kBack:    return osRecordings;
+            case kBack: {  cRecordControl* rc = cRecordControls::GetRecordControl(fileName);
+                           if (rc && rc->InstantId())
+                              return osEnd;
+                           else
+                              return osRecordings;
+                        }
             default:       return osUnknown;
             }
           }
diff -urNad vdr-1.6.0~/menu.h vdr-1.6.0/menu.h
--- vdr-1.6.0~/menu.h	2008-07-25 23:55:15.000000000 +0200
+++ vdr-1.6.0/menu.h	2008-07-25 23:55:15.000000000 +0200
@@ -248,6 +248,7 @@
 public:
   cReplayControl(void);
   virtual ~cReplayControl();
+  void Stop(void);
   virtual cOsdObject *GetInfo(void);
   virtual eOSState ProcessKey(eKeys Key);
   virtual void Show(void);
diff -urNad vdr-1.6.0~/recording.h vdr-1.6.0/recording.h
--- vdr-1.6.0~/recording.h	2008-07-25 23:55:15.000000000 +0200
+++ vdr-1.6.0/recording.h	2008-07-25 23:55:15.000000000 +0200
@@ -196,6 +196,7 @@
 private:
   static const char *command;
 public:
+  static const char *GetCommand(void) { return command; }
   static void SetCommand(const char *Command) { command = Command; }
   static void InvokeCommand(const char *State, const char *RecordingFileName);
   };
